src/back-end/lib/db.ts

```diff
@@ -1921,3 +1921,74 @@ export const readManyCWUSubscribedUsers = tryDb<[Id], User[]>(async (connection,
 
   return valid(await Promise.all(results.map(async raw => await rawUserToUser(connection, raw))));
 });
+
+export const closeCWUOpportunities = tryDb<[], number>(async (connection) => {
+  const now = new Date();
+  return valid(await connection.transaction(async trx => {
+    const lapsedOpportunitiesIds = (await connection<{ id: Id }>('cwuOpportunities as opportunities')
+      .transacting(trx)
+      .join('cwuOpportunityStatuses as statuses', function() {
+        this
+          .on('opportunities.id', '=', 'statuses.opportunity')
+          .andOn('statuses.createdAt', '=',
+            connection.raw('(select max("createdAt") from "cwuOpportunityStatuses" as statuses2 where \
+              statuses2.opportunity = opportunities.id and statuses2.status is not null)'));
+      })
+      .join('cwuOpportunityVersions as versions', function() {
+        this
+          .on('opportunities.id', '=', 'versions.opportunity')
+          .andOn('versions.createdAt', '=',
+            connection.raw('(select max("createdAt") from "cwuOpportunityVersions" as versions2 where \
+              versions2.opportunity = opportunities.id)'));
+      })
+      .where({
+        'statuses.status': CWUOpportunityStatus.Published
+      })
+      .andWhere('versions.proposalDeadline', '<=', now)
+      .select<Array<{ id: Id }>>('opportunities.id'))?.map(result => result.id) || [];
```

Comment:
What an adrenaline rush! 😅

Reply:
:) I might try to abstract some of those joins into functions when I do the refactoring/splitting of `db.ts`.